File: backend/app/schemas/tasks.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class TaskOut(Strict):
    id: UUID
    journey_id: UUID
    property_id: UUID | None
    property_address: str | None
    title: str
    notes: str | None
    assignee_user_id: UUID | None
    assignee_name: str | None
    priority: Literal["low", "normal", "high"]
    due_at: datetime | None
    timezone: str
    reminder_offset_minutes: int | None
    status: Literal["open", "completed"]
    done: bool
    done_at: datetime | None
    row_version: int
    created_at: datetime
    updated_at: datetime


class TaskCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    property_id: UUID | None = None
    title: str = Field(min_length=1, max_length=200)
    notes: str | None = Field(default=None, max_length=4000)
    assignee_user_id: UUID | None = None
    priority: Literal["low", "normal", "high"] = "normal"
    due_at: datetime | None = None
    timezone: str = Field(min_length=1, max_length=64)
    reminder_offset_minutes: int | None = Field(default=None, ge=5, le=10080)


class TaskUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_row_version: int
    property_id: UUID | None = None
    title: str | None = Field(default=None, min_length=1, max_length=200)
    notes: str | None = Field(default=None, max_length=4000)
    assignee_user_id: UUID | None = None
    priority: Literal["low", "normal", "high"] | None = None
    due_at: datetime | None = None
    timezone: str | None = Field(default=None, min_length=1, max_length=64)
    reminder_offset_minutes: int | None = Field(default=None, ge=5, le=10080)
    status: Literal["open", "completed"] | None = None


class MemberOut(BaseModel):
    id: UUID
    display_name: str
    email: str
    role: str


class TaskActivityOut(BaseModel):
    id: UUID
    kind: str
    summary: str
    detail: dict[str, Any]
    created_at: datetime

File: frontend/src/lib/tasks.ts
import { apiFetch } from "./api";

export interface Task {
  id: string; journey_id: string; property_id: string | null; property_address: string | null; title: string; notes: string | null;
  assignee_user_id: string | null; assignee_name: string | null; priority: "low" | "normal" | "high"; due_at: string | null;
  timezone: string; reminder_offset_minutes: number | null; status: "open" | "completed"; done: boolean; done_at: string | null;
  row_version: number; created_at: string; updated_at: string;
}
export interface WorkspaceMember { id: string; display_name: string; email: string; role: string; }
export interface TaskInput { property_id: string | null; title: string; notes: string | null; assignee_user_id: string | null; priority: "low" | "normal" | "high"; due_at: string | null; timezone: string; reminder_offset_minutes: number | null; }

export const taskApi = {
  list: (journeyId: string) => apiFetch<Task[]>(`/journeys/${journeyId}/tasks`),
  members: () => apiFetch<WorkspaceMember[]>("/workspaces/me/members"),
  create: (journeyId: string, body: TaskInput) => apiFetch<Task>(`/journeys/${journeyId}/tasks`, { method: "POST", body: JSON.stringify(body) }),
  update: (journeyId: string, taskId: string, body: Partial<TaskInput> & { expected_row_version: number; status?: "open" | "completed" }) =>
    apiFetch<Task>(`/journeys/${journeyId}/tasks/${taskId}`, { method: "PATCH", body: JSON.stringify(body) }),
  remove: (journeyId: string, taskId: string, expectedRowVersion: number) =>
    apiFetch<void>(`/journeys/${journeyId}/tasks/${taskId}?expected_row_version=${expectedRowVersion}`, { method: "DELETE" }),
  calendarUrl: (journeyId: string, taskId: string) => `${process.env.REACT_APP_BACKEND_URL}/api/journeys/${journeyId}/tasks/${taskId}/calendar.ics`,
};